data_model/derive/src/lib.rs

```diff
@@ -0,0 +1,198 @@
+/// of them.
+#[proc_macro_derive(Filter)]
+pub fn filter_derive(input: TokenStream) -> TokenStream {
+    let ast = syn::parse(input).expect("Failed to parse input TokenStream.");
+    impl_from_filter(&ast)
+}
+
+fn impl_from_filter(ast: &syn::DeriveInput) -> TokenStream {
+    // Omitting attributes as they don't need to be inherited for filters
+    let syn::DeriveInput {
+        vis,
+        ident: event_ident,
+        generics,
+        data,
+        ..
+    } = ast;
+
+    let event_filter_ident = format_ident!("{}Filter", event_ident);
+
+    let filter_ident = format_ident!(
+        "{}Filter",
+        event_ident
+            .to_string()
+            .strip_suffix("Event")
+            .expect("Events should follow the naming format")
+    );
+
+    let event_variants = if let syn::Data::Enum(syn::DataEnum { variants, .. }) = data {
+        variants.iter().collect::<Vec<_>>()
+    } else {
+        panic!("Only enums are supported")
+    };
+
+    let mut filter_variants_id_fields = Vec::new();
+    let mut filter_variants_idents_event_fields = Vec::new();
+
+    let filter_variants_event_fields = event_variants.iter().filter_map(|variant| {
+        let event_filter_variant_name = format_ident!("By{}", variant.ident);
+        if let syn::Fields::Unnamed(ref unnamed) = variant.fields {
+            let variant_type = &unnamed
+                .unnamed
+                .first()
+                .expect("Should have at least one type")
+                .ty;
+            if let syn::Type::Path(path) = variant_type {
+                let var_ty_ident = &path.path.segments[0].ident;
+                // Only `...Event` containing fields are wrapped with `FilterOpt`
+                if var_ty_ident.to_string().contains("Event") {
+                    // Also store just the variant ident for later codegen
+                    filter_variants_idents_event_fields.push(event_filter_variant_name.clone());
+                    let inner_filter_ident = format_ident!(
+                        "{}Filter",
+                        var_ty_ident.to_string().strip_suffix("Event").expect("Checked in outer if"),
+                    );
+                    Some(quote! {
+                        #event_filter_variant_name (crate::prelude::FilterOpt<#inner_filter_ident>)
+                    })
+                } else {
+                    // Default case, make a unit variant with just the ident
+                    filter_variants_id_fields.push(event_filter_variant_name);
+                    None
+                }
+            }
+            else { None }
+        }
+        else { None }
+    }).collect::<Vec<_>>();
+
+    let filter_impl_variant_pairs_event_fields = filter_variants_idents_event_fields
+        .iter()
+        .zip(event_variants.iter())
+        .map(|(filter_variant_ident, event_var)| {
+            let event_var_ident = format_ident!("{}", &event_var.ident);
```

Comment:
OK. I normally find `format.*!("{}"` a little suspicious, always prefer writing what the macro expands to, which here is actually probably just equivalent to a `clone`.